### PR TITLE
get asset from BE #113

### PR DESCRIPTION
**Pull request by Pamela454**
## Dev Summary

[Mocks](link)

On Asset.tsx component,  wrote newAsset to fetch data from backend passing in id param. 

On asset.entity.ts, added column for imgUrls that holds and array with a default value of ["0"]. If a default value is not explicitly stated then it will return undefined and cause an error on the frontend when .filter and .map are called. 

The Asset component displays asset.categories. This data is currently hardcoded. 

## Test Plan
So far added data to database using curl requests. Using console.log confirmed json response on frontend. 

repro steps:

1. `npm run start:dev`
2. Seed data:
 ```
      curl -X POST localhost:3001/api/users \
      -H "Content-Type: application/json" \
      -d '{ "first_name": "test", "last_name": "user", "password": "password", "email": "test@test.com" }'

      curl -X POST localhost:3001/api/users \
     -H "Content-Type: application/json" \
     -d '{ "first_name": "test", "last_name": "user2", "password": "password", "email": "test2@test.com" }'

      curl -X POST localhost:3001/api/assets \
     -H "Content-Type: application/json" \
     -d '{"id": "1",  "title": "thing 1", "description": "its a thing", "type": "donation", "quantity": "1", "imgUrls":     
      "https://optinmonster.com/wp-content/uploads/2019/09/nonprofit-newsletter.png,   
      https://s3.amazonaws.com/mentoring.redesign/s3fs-public/styles/aspect_ratio__4_3/public/nonprofit_biz_tool.jpg",
      "poster": { "first_name":"test", "last_name":"user", "email":"test@test.com", "id": "1" }  }'
```
                                                                                                                                             
4. localhost:3000/assets/1 
5. Expect asset images and asset information to show on page. 3. 

## Resources
information on storing images in an application 
https://wanago.io/2020/08/03/api-nestjs-uploading-public-files-to-amazon-s3/
https://www.reddit.com/r/Nestjs_framework/comments/mtml48/storing_images_best_practice/

Screenshots of asset with and without images in imgUrls.

Screen Shot 2022-04-07 at 10 47 41 AM
Screen Shot 2022-04-07 at 10 48 04 AM

### CODE REVIEW

File: server/src/assets/entities/asset.entity.ts [hunk 1]
```diff
@@ -32,6 +32,10 @@ export class Asset {
   @Column()
   quantity: number;
 
+  @Column('text', { array: true, default: '{}' })
```
Comment (jd2rogers2) [inline server/src/assets/entities/asset.entity.ts:35]: is this `default: '{}'` really how we set the default to an empty array?

File: client/src/features/action/assets/Asset.tsx [hunk 2]
```diff
@@ -204,11 +207,11 @@ function Asset(): JSX.Element {
             {asset.title}
           </Typography>
           <Typography className={classes.subText} variant="subtitle1">
-            {asset.categories.join(', ')}
+            {/*{asset.categories.join(', ')} */}
           </Typography>
           <Typography className={classes.subText} variant="subtitle1">
             {asset.organization ? '' : 'Posted By:'}
-            {asset.postedBy.firstName}
+            {/*{asset.postedBy.firstName} */}
```
Comment (jd2rogers2) [inline client/src/features/action/assets/Asset.tsx:214]: in asset.entity change to `@ManyToOne(() => User, (user) => user.assets, { eager: true })`
and then you should get the `poster` property with a user object as value on client side
will have to change react Asset type to have `poster` not `postedBy`

File: client/src/features/action/assets/Assets.tsx [hunk 3]
```diff
@@ -80,6 +80,9 @@ function Assets(): JSX.Element {
 
   React.useEffect(() => {
     // fetch assets with querySearchText
+    fetch('http://localhost:3001/api/assets')
+      .then((resp) => resp.json())
+      .then((data) => console.log(data));
```
Comment (jd2rogers2) [inline client/src/features/action/assets/Assets.tsx:85]: remove, not within the scope of this PR/ticket

**Comment by jd2rogers2** [comment]
> The Asset component displays asset.categories. I'm not sure if this is supposed to be data held in the asset table? The Asset component is also supposed to display asset.postedBy.firstName. I'm not seeing a relationship in the JSON data that I am getting back. I'm not sure if the join column should be called user_id instead of poster_id in the asset.entity file?

the asset categories relationship is not set up on the BE yet. if you can, i'd say you could hardcode some data in there somehow. lmk if you need help with this hack

re: getting user name on client - left inline comment

> ## Test Plan
> So far added data to database using curl requests. Using console.log confirmed json response on frontend.
can you include the curl request in a snippet in the PR summary/description plz

File: server/src/assets/entities/asset.entity.ts (hunk 1 above)
Comment (Pamela454) [reply server/src/assets/entities/asset.entity.ts:35]: I had seen examples of it here:
https://stackoverflow.com/questions/57611633/typeorm-array-is-not-supported-in-postgres

But I think that empty object will return undefined unless it includes something in it. 

However, I changed it to:

`@Column('text', { array: true, default:["0, 0, 0"] })
    imgUrls: string[];`

This seems to work. 


I guess you can also return an empty array using hooks in typeORM?https://www.darraghoriordan.com/2021/10/19/populating-empty-arrays-type-orm/

File: client/src/features/action/assets/Asset.tsx [hunk 4]
```diff
@@ -135,6 +136,8 @@ function Asset(): JSX.Element {
     return <>asset not found</>;
   }
 
+  console.log(asset);
```
[the remark this answers is not in the document]
Comment (Pamela454) [reply client/src/features/action/assets/Asset.tsx:139]: Deleted

File: client/src/features/action/assets/Asset.tsx (hunk 2 above)
Comment (Pamela454) [reply client/src/features/action/assets/Asset.tsx:214]: Changed and is working now.

Comment (Pamela454) [reply client/src/features/action/assets/Asset.tsx:214]: I changed to poster in Asset type and needed to change first name in User type to first_name as well. I also changed it in several inbox directory files.

**Review by jd2rogers2 — COMMENTED** [review]
one super little thing
then 🚢🚢🚢🚢🚢🚢🚢🚢🚢

**Comment by Pamela454** [comment]
I saw your feedback on the PR. I tried running the command git checkout upstream main -- README.md that you wrote and got an error "fatal: only one reference expected, 2 given." Running git checkout upstream/main -- README.md did work. Looks like that file change is no longer showing up.

File: README.md [hunk 5]
```diff
@@ -40,7 +40,7 @@
         * stop service: `brew services stop postgresql`
         * restart service: `brew services restart postgresql`
     * once postgres is running, try `psql`, it will say `Password for user USERNAME`, try some usual passwords
-        * if you get in:
+        * if you get in:
```
[the remark this answers is not in the document]
Comment (Pamela454) [reply README.md]: I tried running the command git checkout upstream main -- README.md that you wrote and got an error "fatal: only one reference expected, 2 given." Running git checkout upstream/main -- README.md did work. Looks like that file change is no longer showing up.

**Review by jd2rogers2 — APPROVED** [review]
woops, meant to approve last review when i said 🚢❗🚢❗🚢❗🚢❗🚢❗🚢❗🚢❗🚢❗🚢❗

**Pull request merged**